src/com/reason/ORCompilerManager.java

```diff
@@ -0,0 +1,103 @@
+    }
+
+    public Optional<Compiler> getCompiler(CliType cliType) {
+        return getCompiler(cliType.getCompilerType());
+    }
+
+    public Optional<Compiler> getCompiler(CompilerType compilerType) {
+        Compiler compiler = ServiceManager.getService(project, getCompilerClass(compilerType));
+        if (compiler.isConfigured(project)) {
+            return Optional.of(compiler);
+        }
+        return Optional.empty();
+    }
+
+    public Optional<Compiler> getCompiler(VirtualFile editorFile) {
+        FileType fileType = editorFile.getFileType();
+        if (!Compiler.isSupportedFileType(fileType)) {
+            return Optional.empty();
+        }
+        return traverseAncestorsForCompiler(editorFile.getParent(), new HashMap<>());
+    }
+
+    private Optional<Compiler> traverseAncestorsForCompiler(VirtualFile currentDir,
+            HashMap<String, VirtualFile> visited) {
+        // hit filesystem root, give up
+        if (currentDir == null) {
+            return Optional.empty();
+        }
+        // we've already visited this directory, must have hit a symlink
+        if (visited.get(currentDir.getPath()) != null) {
+            return Optional.empty();
+        }
+        visited.put(currentDir.getPath(), currentDir);
+        // look for a compiler configuration file
+        for (VirtualFile child : currentDir.getChildren()) {
+            if (EsyPackageJson.isEsyPackageJson(child)) {
+                Optional<Compiler> compiler = getCompiler(CompilerType.ESY);
+                if (compiler.isPresent()) {
+                    return compiler;
+                }
+            }
+            if (DuneFileType.isDuneFile(child)) {
+                // will be empty if dune isn't configured, might be an esy project
+                Optional<Compiler> compiler = getCompiler(CompilerType.DUNE);
+                if (compiler.isPresent()) {
+                    return compiler;
+                }
+            }
+            if (BsConfigJsonFileType.isBsConfigFile(child)) {
+                Optional<Compiler> compiler = getCompiler(CompilerType.BS);
+                if (compiler.isPresent()) {
+                    return compiler;
+                }
+            }
+        }
+        // we just checked the project root, we're done
+        if (currentDir.getPath().equals(project.getBasePath())) {
+            return Optional.empty();
+        }
+        // move up a directory and try again
+        return traverseAncestorsForCompiler(currentDir.getParent(), visited);
+    }
+
+    private static Class<? extends Compiler> getCompilerClass(CompilerType compilerType) {
+        switch (compilerType) {
+            case BS:
+                return BsCompiler.class;
+            case DUNE:
+                return DuneCompiler.class;
+            case ESY:
+                return EsyCompiler.class;
+            default:
+                throw new RuntimeException("Unsupported compiler type. Type = " + compilerType);
```

Comment:
the idea was that there is always a valid compiler: the dummy compiler. either it is the case, and we don't have exception, nor optional in signatures - because there is a dummy compiler - , or we use exceptions.
My idea is that we should be less intrusive in the user wokrflow and avoid exceptions when possible (like #234).

Reply:
So that default case should never be hit. I needed to add it to make the compiler happy.

Reply:
you're right.
what about a different message that doesn't look like a real exception ? something like "Default case is unreachable" (with better english :) )

Reply:
good idea. changed the default case to log an error and return the BuckleScript compiler